---
title: "Gene Ontology (GO) Gene Set Enrichment Analysis using the package *gsea*"
author: "Amir Levine"
date: "2018-11-06"
output: md_document
---

# Pre-loading

Load packages.
```{r}
source("https://bioconductor.org/biocLite.R")
if (!require(biomaRt)) {BiocManager::install("biomaRt"); library(biomaRt)}
if (!require(org.Ce.eg.db)) {BiocManager::install("org.Ce.eg.db"); library(org.Ce.eg.db)}
if (!require(fgsea)) {BiocManager::install("fgsea"); library(fgsea)}
```

Load two pre-processed datasets:
1. `DEG` is a list of sets of gene IDs that were differentially regulated by *anc-1* RNAi.
2. `BOTH` is a tibble that contains the RNA-seq gene expression information and statistics for the two strains.
```{r}
load("data/2018-11-27-geneExpression-pval-thresh-0.05.rdata")
```

Load biomart object of *C. elegans* GO annotations.
```{r}
load("data/2018-11-27-GO-annotations.Rdata")
```


# Ranked Gene list
The ranked gene-list for the analysis needs to be ordered by Fold-change.
The list is sorted in a descending order. 
To avoid some combination of these numerical values in the two strains, I'll analyze the gene-lists from each strain independently. **Split this into 2 analyses**.
Remove NA values from adjusted *p*-value.
Return a vector of the Gene IDs for all the identified genes.
**Consider using just the *p*-values to rank the list**
```{r Ranked Gene List}
#strain = "N2"
strain = "AM140"

geneList <- BOTH %>% 
  dplyr::filter(!is.na(padj)) %>%
  dplyr::mutate(rank = log2FoldChange) %>%
  dplyr::filter((!!strain) == strain) %>%
  dplyr::select(Gene_stable_ID,rank)

rankedList <- as.vector(geneList$rank) 
names(rankedList) <- geneList$Gene_stable_ID
rankedList <- sort(rankedList, decreasing = TRUE)
```

# GSEA: GOSlim Terms
View possible GO terms annotations are:
```{r}
levels(factor(goslim_annotation$namespace_1003))
```
```
[1] ""                   "biological_process" "cellular_component" "eco"               
[5] "EFO"                "molecular_function"
```

## Process GOSlim Terms
Extract GO annotations that are under *Biological Process*.  
Generate **Gene-Sets**, a list object of Ensembl Gene IDs vectors that correspond to a unique GOSlim term.
```{r GO BP annotations}
goslim_annotation_bp <- goslim_annotation %>%
  filter(namespace_1003 == "biological_process", go_linkage_type != "IEA")

goslim_pathway_sets <- split(goslim_annotation_bp$ensembl_gene_id,
                             paste(goslim_annotation_bp$goslim_goa_accession,
                                   goslim_annotation_bp$goslim_goa_description))
goslim_pathway_sets <- lapply(goslim_pathway_sets, unique)
```

## GSEA
Perform GSEA, focusing on medium-sized gene-sets (with 15-500 members), 10,000 permutations.
Pring the enriched (*p*-value < 0.05) gene-sets.
```{r}
set.seed(159159)
fgseaRes <- fgsea(pathways = goslim_pathway_sets, 
                  stats = rankedList,
                  minSize=15,
                  maxSize=500,
                  nperm=10000)
head(fgseaRes[order(pval),][padj < 0.05],15)
```


# GSEA: GO terms

## Process GO Terms
Extract GO annotations that are under *Biological Process*
```{r GO BP annotations}
go_annotation_bp <- go_annotation %>%
  filter(namespace_1003 == "biological_process", go_linkage_type != "IEA")
```

Generate **Gene-Sets**, a list object of Ensembl Gene IDs vectors that correspond to a unique GOSlim term.
```{r Generate **Gene-Sets**}
go_pathway_sets <- split(go_annotation_bp$ensembl_gene_id,
                         paste(go_annotation_bp$go_id,
                               go_annotation_bp$name_1006))
go_pathway_sets <- lapply(go_pathway_sets, unique)
```

## GSEA
Perform GSEA, focusing on medium-sized gene-sets (with 15-500 members), 10,000 permutations.
Pring the enriched (*p*-value < 0.05) gene-sets.
```{r}
set.seed(159159)
fgseaRes <- fgsea(pathways = go_pathway_sets, 
                  stats = rankedList,
                  minSize=15,
                  maxSize=500,
                  nperm=10000)
head(fgseaRes[order(pval),][padj < 0.05],15)
```
...



# Convert IDs
Convert gene IDs from *Ensembl* to *Entrez* using *BioMart*, a required inout ID for KEGG enrichment analysis in the package *clusterProfilter*. This is not a requirement for Gene Ontology enrichment analyses using the same package.
```{r}
ENTREZ.ID <- bitr(goi, # All regulated genes
                  fromType="WORMBASE", # From Stable Gene ID
                  toType="ENTREZID", # To Entrez ID
                  OrgDb="org.Ce.eg.db", # C. elegans annotation data.
                  drop=TRUE)
```

The above code returns a warning regarding 1-to-many mappings and 0.13% of input genes had no match. Should perhaps change the algorithm to one that uses *Ensembl* gene IDs.
```
# 'select()' returned 1:many mapping between keys and columns
# Warning message:
# In bitr(DEG$UNION.reg, fromType = "WORMBASE", toType = "ENTREZID",  :
#   0.13% of input gene IDs are fail to map...
```


# KEGG Overrepresentation Analysis
...Need to perform...

kk <- enrichKEGG(gene         = ids[['ENTREZID']],
                 keyType      = 'ncbi-geneid',
                 organism     = 'cel', 
                 pvalueCutoff = 0.05)
head(kk, n=15)

## plotting and pathway visualization ##

# barplot(kk,
#         showCategory=12,
#         drop=TRUE,
#         colorBy = 'p.adjust'
#         )
#
#dotplot(kk,showCategory=12)

##Visualize the pathways
# browseKEGG(kk,"cel04120")


## plotting by ggplot2

kk <- as.data.frame(kk)

factor(kk$p.adjust,levels = names(table(kk$p.adjust))[order(kk$p.adjust)])
order(kk$p.adjust)
ggplot(kk, aes(x = reorder(Description,p.adjust), y = Count)) +
  geom_bar(aes(fill = p.adjust),stat = 'identity') + 
  coord_flip() + 
  theme_classic() + 
  scale_fill_gradient(name = "P value",low = "red3",high = "orange") + 
  #ggtitle("KEGG pathways") + 
  ylab("Gene count") + 
  xlab("") + 
  theme(text = element_text(size=7))
ggsave("results\\KEGG_OVERREP_BOTH.DOWN.wmf", height = 4.5, width = 10, units = "cm")

## export data to dable
write.csv(as.data.frame(kk),'results\\KEGG_OVERREP_BOTH.DOWN.csv')


# GO Gene Set Enrichment Analysis (GSEA) - *clusterProfiler*
...Need to perform...

ego3 <- gseGO(geneList     = geneList,
              keyType      = "WORMBASE",
              OrgDb        = org.Ce.eg.db,
              ont          = "BP",
              nPerm        = 100,
              minGSSize    = 10,
              maxGSSize    = 500,
              pvalueCutoff = 0.05,
              verbose      = TRUE)
              
# KEGG Gene Set Enrichment Analysis (GSEA) - *clusterProfiler*
...Need to perform...

kk2 <- gseKEGG(geneList     = gsea,
                keyType      = 'ncbi-geneid',
                organism     = 'cel',
                nPerm        = 1000,
                minGSSize    = 40,
                pvalueCutoff = 0.05,
                verbose      = TRUE)
head(kk2)
```
## Results, due to minGSSize    = 40, other values remove some of the results ##
# ID                    Description setSize enrichmentScore       NES      pvalue   p.adjust    qvalues rank
# cel03010 cel03010                       Ribosome     125       0.6713444  2.237956 0.001106195 0.03429204 0.02794597 2744
# cel03015 cel03015      mRNA surveillance pathway      69      -0.5240880 -1.999406 0.005952381 0.04696970 0.03827751 3162
# cel00240 cel00240          Pyrimidine metabolism      72      -0.3799745 -1.447325 0.006289308 0.04696970 0.03827751 4231
# cel04120 cel04120 Ubiquitin mediated proteolysis      80      -0.4375823 -1.713697 0.006535948 0.04696970 0.03827751 3425
# cel03013 cel03013                  RNA transport     114      -0.3601193 -1.526163 0.008547009 0.04696970 0.03827751 4167
# cel03040 cel03040                    Spliceosome     111      -0.5156142 -2.145414 0.009090909 0.04696970 0.03827751 3558
# leading_edge
# cel03010 tags=64%, list=23%, signal=50%
# cel03015 tags=51%, list=26%, signal=38%
# cel00240 tags=54%, list=35%, signal=35%
# cel04120 tags=48%, list=28%, signal=34%
# cel03013 tags=50%, list=34%, signal=33%
# cel03040 tags=65%, list=29%, signal=47%
```


sessionInfo()
```
# R version 3.5.1 (2018-07-02)
# Platform: i386-w64-mingw32/i386 (32-bit)
# Running under: Windows >= 8 x64 (build 9200)
# 
# Matrix products: default
# 
# locale:
#   [1] LC_COLLATE=English_Israel.1252  LC_CTYPE=English_Israel.1252    LC_MONETARY=English_Israel.1252
# [4] LC_NUMERIC=C                    LC_TIME=English_Israel.1252    
# 
# attached base packages:
#   [1] stats4    parallel  grid      stats     graphics  grDevices utils     datasets  methods  
# [10] base     
# 
# other attached packages:
#   [1] org.Ce.eg.db_3.6.0    AnnotationDbi_1.42.1  Biobase_2.40.0        clusterProfiler_3.8.1
# [5] VennDiagram_1.6.20    futile.logger_1.4.3   bindrcpp_0.2.2        PWMEnrich_4.16.0     
# [9] Biostrings_2.48.0     XVector_0.20.0        IRanges_2.14.10       S4Vectors_0.18.3     
# [13] BiocGenerics_0.26.0   forcats_0.3.0         stringr_1.3.1         dplyr_0.7.6          
# [17] purrr_0.2.5           readr_1.1.1           tidyr_0.8.1           tibble_1.4.2         
# [21] ggplot2_3.0.0         tidyverse_1.2.1       BiocInstaller_1.30.0 
# 
# loaded via a namespace (and not attached):
#   [1] gridExtra_2.3        GOSemSim_2.6.2       stringi_1.1.7        evaluate_0.11       
# [5] memoise_1.1.0        haven_1.1.2          lattice_0.20-35      cli_1.0.0           
# [9] RSQLite_2.1.1        fansi_0.3.0          evd_2.3-3            blob_1.1.1          
# [13] DBI_1.0.0            data.table_1.11.4    rstudioapi_0.7       bindr_0.1.1         
# [17] units_0.6-1          nlme_3.1-137         zlibbioc_1.26.0      qvalue_2.12.0       
# [21] ggrepel_0.8.0        UpSetR_1.3.3         rprojroot_1.3-2      tools_3.5.1         
# [25] magrittr_1.5         Rcpp_0.12.18         xml2_1.2.0           gdata_2.18.0        
# [29] readxl_1.1.0         httr_1.3.1           rmarkdown_1.10       assertthat_0.2.0    
# [33] R6_2.2.2             fastmatch_1.1-0      munsell_0.5.0        cellranger_1.1.0    
# [37] gtools_3.8.1         digest_0.6.16        splines_3.5.1        cowplot_0.9.3       
# [41] DOSE_3.6.1           colorspace_1.3-2     ggraph_1.0.2         pkgconfig_2.0.2     
# [45] pillar_1.3.0         GO.db_3.6.0          formatR_1.5          tweenr_1.0.0        
# [49] rvcheck_0.1.1        plyr_1.8.4           farver_1.0           gtable_0.2.0        
# [53] tidyselect_0.2.4     rvest_0.3.2          reshape2_1.4.3       knitr_1.20          
# [57] viridisLite_0.3.0    futile.options_1.0.1 rlang_0.2.2          broom_0.5.0         
# [61] glue_1.3.0           Matrix_1.2-14        backports_1.1.2      scales_1.0.0        
# [65] ggridges_0.5.1       lubridate_1.7.4      lambda.r_1.2.3       modelr_0.1.2        
# [69] igraph_1.2.2         enrichplot_1.0.2     ggforce_0.1.3        fgsea_1.6.0         
# [73] hms_0.4.2            labeling_0.3         htmltools_0.3.6      yaml_2.2.0          
# [77] lazyeval_0.2.1       utf8_1.1.4           crayon_1.3.4         withr_2.1.2         
# [81] MASS_7.3-50          BiocParallel_1.14.1  bit64_0.9-7          seqLogo_1.46.0      
# [85] DO.db_2.9            viridis_0.5.1        jsonlite_1.5         compiler_3.5.1      
# [89] bit_1.1-14    
```
