---
title: "TBX-39/40 and TBX-38/43 binding motifs genome wide"
author: "Amir Levine"
date: "2019-02-04"
output: md_document
---

Load packages and datasets.
```{r packages}
if (!require(tidyverse)) {install.packages("tidyverse"); library(tidyverse)}
if (!require(Biostrings)) {BiocManager::install("Biostrings"); library(Biostrings)} # Aligning motifs
if (!require(eulerr)) {BiocManager::install("eulerr"); library(eulerr)} # Plotting Euler Diagrams
```

Obtain the 500 bp upstream, and 100 bp downstream, promoter regions of all the protein-coding genes in the *C. elegans* genome. The data was originally downloaded using the script `1.9-Celegans-promoters-BioMaRt.R`. 
Obtain the PWMs from script `2.1-Celegans-TFs-PWM-clean.R`.
```{r load-datasets}
load("data/Celegans-promoters.rdata") # Get promoters
load("experiments/2018-09-02-TFs-motifs/data/TFs-PWM.rdata") # Get PWMs
# load("data/2018-11-27-geneExpression-pval-thresh-0.05.rdata") # Get CEL-seq processed data of anc-1 RNAi, p<0.05
```

```{r dir-structure}
dir <- "experiments/2019-02-05-tbx-presence/"
dir.create(dir, showWarnings = FALSE)
dir.create(paste0(dir,"results"), showWarnings = FALSE)
dir.create(paste0(dir,"data"), showWarnings = FALSE)
```

Subset PWMs that belong to T-box TFs. Then, convernt PWMs to Position Count Matrix (PCM), to serve as input to the `matchPWM` function.
```{r PWM-to-PCM}
PWM <- PWM[grep("tbx-",names(PWM))]
PCM <- lapply(PWM,"*",100)
PCM <- lapply(PCM,round)
```

Scan all *C. elegans* coding gene promoter regions (+500/-100 bp of TSS) for each T-box transcription factor binding motifs.
```{r scan-motifs}
pwm.tbx33.hits <- sapply(prs,function(pseq)
  matchPWM(PCM$`tbx-33`, pseq, min.score="90%", with.score = TRUE))
pwm.tbx38.hits <- sapply(prs,function(pseq)
  matchPWM(PCM$`tbx-38`, pseq, min.score="90%", with.score = TRUE))
pwm.tbx39.hits <- sapply(prs,function(pseq)
  matchPWM(PCM$`tbx-39`, pseq, min.score="90%", with.score = TRUE))
pwm.tbx40.hits <- sapply(prs,function(pseq)
  matchPWM(PCM$`tbx-40`, pseq, min.score="90%", with.score = TRUE))
pwm.tbx43.hits <- sapply(prs,function(pseq)
  matchPWM(PCM$`tbx-43`, pseq, min.score="90%", with.score = TRUE))
```

Save intermediate scan results for later usage.
```{r scan-save}
save(pwm.tbx33.hits,
     pwm.tbx38.hits,
     pwm.tbx39.hits,
     pwm.tbx40.hits,
     pwm.tbx43.hits,
     file = paste0(dir,"data/2019-02-05-promoters-scan-90p.Rdata"))
```


```{r}
load(paste0(dir,"data/2019-02-05-promoters-scan-90p.Rdata"), verbose = TRUE)
```

Change the strucutre from one that details the position of each motifs on each sequence, to a count of positions, and merge to a single matrix, where the rows are genes, and columns are T-box binding sites.
```{r ignore-positions}
counts <- cbind(sapply(pwm.tbx33.hits, length),
                sapply(pwm.tbx38.hits, length),
                sapply(pwm.tbx39.hits, length),
                sapply(pwm.tbx40.hits, length),
                sapply(pwm.tbx43.hits, length))
colnames(counts) <- c("TBX-33","TBX-38","TBX-39","TBX-40","TBX-43")
```

Consider every promoter that possesses at least one motif as `TRUE`.
```{r counts-to-presence}
motifs <- counts > 0
```

```{r print-motifs}
colSums(motifs)
```

```{r euler-plot}
## TBX39-TBX40-TBX3843

motifs2 <- motifs[,3:4]
motifs2 <- cbind(motifs2,motifs[,2]|motifs[,5])
colnames(motifs2) <- c("TBX-39","TBX-40","TBX-38/TBX-43")


vd <- euler(motifs2, shape = "ellipse") # Generate Euler
error_plot(vd) # Errors in plotting estimates.

cairo_pdf(
  filename = paste0(dir,"results/2019-02-06-euler-TBX-four.pdf"),
  width = 1.5, height = 1.5)
    
plot(
  vd, 
  lwd = 1,
  fills = list(fill = c("peachpuff","mistyrose","lightgrey")),
  labels = F, quantities = F
  #labels = list(fontsize = 7),
  #quantities = list(fontsize = 7)
)

dev.off()

### TBX38-TBX43
vd <- euler(motifs[,c(2,5)], shape = "ellipse") # Generate Euler
error_plot(vd) # Errors in plotting estimates.

cairo_pdf(
  filename = paste0(dir,"results/2019-02-06-euler-TBX38.43.pdf"),
  width = .75, height = .75)
    
plot(
  vd, 
  lwd = 1,
  fills = list(fill = c("darkgrey","white")),
  labels = F, quantities = F
  #labels = list(fontsize = 7),
  #quantities = list(fontsize = 7)
)

dev.off()
```

```{r session-info}
sessionInfo()
```
```
R version 3.5.1 (2018-07-02)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows >= 8 x64 (build 9200)

Matrix products: default

locale:
[1] LC_COLLATE=English_United States.1252  LC_CTYPE=English_United States.1252   
[3] LC_MONETARY=English_United States.1252 LC_NUMERIC=C                          
[5] LC_TIME=English_United States.1252    

attached base packages:
 [1] stats4    parallel  grid      stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] PWMEnrich_4.18.0    Biostrings_2.50.0   XVector_0.22.0      IRanges_2.16.0      S4Vectors_0.20.0   
[6] BiocGenerics_0.28.0 eulerr_5.1.0       

loaded via a namespace (and not attached):
 [1] Rcpp_0.12.19       gtools_3.8.1       zlibbioc_1.28.0    gdata_2.18.0       tools_3.5.1       
 [6] polyclip_1.9-1     polylabelr_0.1.0   yaml_2.2.0         compiler_3.5.1     seqLogo_1.48.0    
[11] BiocManager_1.30.3 evd_2.3-3          knitr_1.20        
```
