---
title: "tbox motif enrichment"
output: html_document
---

## additional sources
* http://master.bioconductor.org/packages/release/workflows/vignettes/generegulation/inst/doc/generegulation.html
* https://www.bioconductor.org/packages/release/bioc/vignettes/PWMEnrich/inst/doc/PWMEnrich.pdf
* https://cran.r-project.org/web/packages/rtfbs/vignettes/vignette.pdf
* http://bioconductor.org/packages/release/bioc/vignettes/TFBSTools/inst/doc/TFBSTools.html

rm(list = ls())

## load clean CEL-seq gene lists
```{r}
load("data\\CELseq_combined.rdata")
```

## load bioconductor and packages
```{r}
source("https://bioconductor.org/biocLite.R")

if (!require(tidyverse)) {install.packages("tidyverse"); library(tidyverse)}

# biocLite("org.Ce.eg.db") ; library(org.Ce.eg.db) #Genome wide annotation for Worm, primarily based on mapping using Entrez Gene identifiers.
# biocLite("BiocGenerics") ;library(BiocGenerics)
# biocLite("AnnotationHub") ; library(AnnotationHub)
```

## Obtain promoters of C. elegans genes (shortcut)
Load promoters of C. elegans genes. Data generated by "Celegans-promoters.R".
```{r}
load("data/Celegans-promoters.rdata")
```

# T-box transcriptions factors

## Obtain motifs and define background set of sequences (shortcut)
*Can skip the next two section by loading the final output - the background sequences*
```{r}
load("experiments\\2018-07-11-tbx-motifs\\data\\tbx-motifs-enrichment\\backgound-celegans-proteincoding-tbx.rdata")
```

## Obtain motifs
import the Position Weight Matrices for the T-box genes, which were manually downlaoded from the (CIS-BP Database)[http://cisbp.ccbr.utoronto.ca] (only TF info and PWMs) and  manually copied each matrix to an individual text file.
1. obtain the file paths and read each into a matrix within a list
2. obtrain the file names
3. assign the filenames to the list matrices
4. transpose the matrices, as that is the acceptable input format (rows = bases, columns = positions)
```{r}
motifs.pwm <- lapply(Sys.glob("experiments\\2018-07-11-tbx-motifs\\raw\\CisBP_2018_07_09_3_15_am\\cisBP-PWMs-textfiles\\*.txt"), read.table,header = T,row.names = 1)
TF.names <- list.files("experiments\\2018-07-11-tbx-motifs\\raw\\CisBP_2018_07_09_3_15_am\\cisBP-PWMs-textfiles\\",full.names=F)
names(motifs.pwm) <- TF.names
motifs.pwm <- lapply(motifs.pwm,t)
```

convert PWM to PCM (partially arbitrariy by multiplying by 100), rounding, and converting the matrix type to "integer" (required input to the enrichment calculation algorithm)
```{r}
motifs.pcm <- lapply(motifs.pwm,"*",100)
motifs.pcm <- lapply(motifs.pcm,round)
motifs.pcm <- lapply(motifs.pcm, FUN= function(x) type.convert(x))
# storage.mode(motifs.pcm[[1]]) # verify the matrix is stored as an integer
```

## Define background set of sequences
Using a custom set of background sequences for C. elegans, based specifically on the promoters I defined, and the motifs I've chose.
the sequences are split into 100bp chunks and fitted
As the process was very long, so it's best to save the output background for future sessions
```{r}
install.packages("PWMEnrich"); library(PWMEnrich)
bg.custom <- makeBackground(motifs = motifs.pcm,
                            type = "logn",
                            bg.seq = prs,
                            bg.len=100,
                            bg.source="all promoters split into 100bp chunks")
save(bg.custom,file = "experiments\\2018-07-11-tbx-motifs\\data\\tbx-motifs-enrichment\\backgound-celegans-proteincoding-tbx.rdata")
# load("tbx-motifs-enrichment\\backgound-celegans-proteincoding-tbx.rdata") # to load the background sequences
```


# IIS transcription factors: daf-16, hsf-1, skn-1 + skn-1

## Obtain motifs and define background set of sequences (shortcut)
*Can skip the next two section by loading the final output - the background sequences*
```{r}
load("experiments\\2018-07-11-tbx-motifs\\data\\backgound-celegans-proteincoding-daf16-skn1-hsf1-pha4.rdata")
```

## Obtain motifs
import the Position Weight Matrices for the T-box genes, which were manually downlaoded from the (CIS-BP Database)[http://cisbp.ccbr.utoronto.ca] (only TF info and PWMs) and  manually copied each matrix to an individual text file.
1. obtain the file paths and read each into a matrix within a list
2. obtrain the file names
3. assign the filenames to the list matrices
4. transpose the matrices, as that is the acceptable input format (rows = bases, columns = positions)
```{r}
motifs.pwm <- lapply(Sys.glob("experiments\\2018-07-11-tbx-motifs\\raw\\CisBP_2018_07_11_7_06_am\\cisBP-PWMs-textfiles\\*.txt"), read.table,header = T,row.names = 1)
TF.names <- list.files("experiments\\2018-07-11-tbx-motifs\\raw\\CisBP_2018_07_11_7_06_am\\cisBP-PWMs-textfiles\\",full.names=F)
names(motifs.pwm) <- TF.names
motifs.pwm <- lapply(motifs.pwm,t)
```

convert PWM to PCM (partially arbitrariy by multiplying by 100), rounding, and converting the matrix type to "integer" (required input to the enrichment calculation algorithm)
```{r}
motifs.pcm <- lapply(motifs.pwm,"*",100)
motifs.pcm <- lapply(motifs.pcm,round)
motifs.pcm <- lapply(motifs.pcm, FUN= function(x) type.convert(x))
# storage.mode(motifs.pcm[[1]]) # verify the matrix is stored as an integer
```

## Define background set of sequences
Using a custom set of background sequences for C. elegans, based specifically on the promoters I defined, and the motifs I've chose.
the sequences are split into 100bp chunks and fitted
As the process was very long, so it's best to save the output background for future sessions
```{r}
install.packages("PWMEnrich"); library(PWMEnrich)
bg.custom <- makeBackground(motifs = motifs.pcm,
                            type = "logn",
                            bg.seq = prs,
                            bg.len=100,
                            bg.source="all promoters split into 100bp chunks")
save(bg.custom,file = "experiments\\2018-07-11-tbx-motifs\\data\\backgound-celegans-proteincoding-daf16-skn1-hsf1-pha4.rdata")
# load("backgound-celegans-proteincoding-daf16-skn1-hsf1-pha4.rdata") # to load the background sequences
```


# Perform motif enrichment!

## List all the genes that are significantly regulated by anc-1
* Genes of interest - all genes that are significantly downregulated in AM140 by anc-1
* generate query/primary sequences list (TRUE values, hits from differential expression analysis) and control/background sequences list (FALSE values, all the other protein-coding genes)
```{r}
goi <- both.sets$N2.dn
mRNA.seqs.promoter <- mRNA.seqs.promoter %>% mutate(CELseq.hits = ensembl_gene_id %in% goi)
```

## Motif enrichment scanning
Analysis is based on the bioconductor (PWMEnrich  package)[https://www.bioconductor.org/packages/release/bioc/html/PWMEnrich.html]
```{r results='hide' message=FALSE warning=FALSE include=FALSE }
res = motifEnrichment(prs[mRNA.seqs.promoter$CELseq.hits], bg.custom)
```

## Generate a report
```{r}
report <- groupReport(res)
pdf(file = "results\\tbx-motifs-enrichment\\PWMreport.pdf",width=8,height=4)
plot(report, fontsize=7, id.fontsize=5)
dev.off()
```



# Plotting a heatmap of the p-value for each 

## Create p-values matrix
Obtain the p-values of motifs per sequence.
* rows correspond to the different input sequences and the columns correspond to motifs.
* sequence.nobg contains the raw affinity scores.
* sequence.bg contains the corresponding P-values.
```{r}
pvals.mat <- res$sequence.bg
```

Define the names of the t-box proteins to be used as the x-axis label.
```{r}
colnames(pvals.mat) <- c("TBX-33","TBX-38","TBX-39","TBX-40","TBX-43")
```

Change the gene IDs to common gene names
```{r}
rownames(pvals.mat) <- mRNA.ids$external_gene_name[which(mRNA.ids$ensembl_gene_id %in% rownames(pvals.mat))]
```

## Change p-value to boolean, which are TRUE if statistically significant and filter-out genes w/ no motif
```{r}
pvals.mat.sig <- (pvals.mat < 0.05) * 1
pvals.mat.sig <- pvals.mat.sig[rowSums(pvals.mat.sig) > 0,]
round(sum(pvals.mat.sig[,1]) / nrow(pvals.mat) * 100)
round(sum(pvals.mat.sig[,2]) / nrow(pvals.mat) * 100)
round(sum(pvals.mat.sig[,3]) / nrow(pvals.mat) * 100)
round(sum(pvals.mat.sig[,4]) / nrow(pvals.mat) * 100)
round(sum(pvals.mat.sig[,5]) / nrow(pvals.mat) * 100)

```

## (for grant propoal) 
```{r}
row_dend = hclust(dist(pval.mat.sig)) # row clustering
col_dend = hclust(dist(t(pval.mat.sig))) # column clustering

pdf("tbx_heatmap_up_grant.pdf",width=4,height=10)
Heatmap(pval.mat.sig, 
        name = "Motif enrichment p-value", #title of legend
        column_title = "T-BOX motif enrichment", 
        row_title = "Genes upregulated by anc-1 RNAi",
        column_names_side = "top",
        row_names_gp = gpar(fontsize = 7), # Text size for row names
        column_names_gp = gpar(fontsize = 7), # Text size for column names
        cluster_rows = color_branches(row_dend, k = 2),
        #km = 8 ,# To split the dendrogram using k-means, divide into 4 groups. Must set.seed
        cluster_columns = color_branches(col_dend, k = 3),
        col = mycols)

dev.off()

dev.print(png, 'filename.png')
dev.copy2pdf('filename.pdf',out.type = "pdf")
dev.off()

```




## Create p-values data-frame
Add 'gene name' variable to p-value dataframe, and make long data
```{r}
pvals.df.long <- pvals.mat.sig %>%
  as_tibble(rownames = "external_gene_name") %>% 
  gather(TF.motifs, p.value, -external_gene_name)
```

## Is there a correlation between any two motifs?
```{r}
cor <- cor(pvals.mat, method = "pearson")
if (!require(Hmisc)) {install.packages("Hmisc"); library(Hmisc)}
cor2 <- rcorr(pvals.mat)
```

The function rcorr() [in Hmisc package] can be used to compute the significance levels for pearson and spearman correlations. It returns both the correlation coefficients and the p-value of the correlation for all possible pairs of columns in the data table.
```{r}
if (!require(corrplot)) {install.packages("corrplot"); library(corrplot)}

corrplot(cor, type = "upper", order = "hclust", 
         tl.col = "black", tl.srt = 45)
```

## Plot heatmap of p-values ver2
based on (Heatmap - Static and Interactive: Absolute Guide
)[http://www.sthda.com/english/articles/28-hierarchical-clustering-essentials/93-heatmap-static-and-interactive-absolute-guide/]
```{r}

if (!require(dendextend)) {install.packages("dendextend"); library("dendextend")}
               
row_dend = hclust(dist(pvals.mat.sig)) # row clustering
col_dend = hclust(dist(t(pvals.mat.sig))) # column clustering

if (!require(ComplexHeatmap)) {biocLite("ComplexHeatmap"); library("ComplexHeatmap")}
Heatmap(pvals.mat.sig,
        name = "Motif presence (p-value < 0.05)", # title of legend
        row_title = "Genes downregulated by anc-1 RNAi",
        column_names_side = "top",
        row_names_gp = gpar(fontsize = 6), # Text size for row names
        cluster_rows = color_branches(row_dend, k = 8),
        cluster_columns = color_branches(col_dend, k = 3),
        col = c("grey","blue")) 

require("pheatmap")
pheatmap(pvals.mat.sig, 
         cutree_rows = 8, 
         color = c("white","red"),
         cellwidth = 5,
         cellheight = 5)


```

## Plot heatmap of p-values
```{r}
res.pvals %>% 
  ggplot(aes(x = TF.motifs, y = external_gene_name)) +
  geom_tile(aes(fill = p.value)) +
  scale_fill_viridis() +
  ylab("Genes ") +
  xlab("T-box binding motifs") +
  theme_bw() + 
  theme(legend.title = element_text(size = 10),
        legend.text = element_text(size = 12),
        plot.title = element_text(size=16),
        axis.title=element_text(size=14,face="bold"),
        axis.text.y=element_text(size=7),
        axis.text.x=element_text(size=14)) +
  labs(fill = "p-value")
```

based on (Heatmap - Static and Interactive: Absolute Guide
)[http://www.sthda.com/english/articles/28-hierarchical-clustering-essentials/93-heatmap-static-and-interactive-absolute-guide/]
```{r}
df <- pvals.mat
clustering_distance_rows, clustering_distance_columns: metric for clustering: “euclidean”, “maximum”, “manhattan”, “canberra”, “binary”, “minkowski”, “pearson”, “spearman”, “kendall”)
```